cobc/gentable.c

```diff
@@ -0,0 +1,145 @@
+   as published by the Free Software Foundation, either version 3 of the
+   License, or (at your option) any later version.
+
+   GnuCOBOL is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+   GNU General Public License for more details.
+
+   You should have received a copy of the GNU General Public License
+   along with GnuCOBOL.  If not, see <https://www.gnu.org/licenses/>.
+*/
+
+#include "config.h"
+#include "cobc.h"
+#include "tree.h" /* for cb_note */
+
+#include <stdio.h>
+
+#ifdef HAVE_ICONV
+
+#include <iconv.h>
+#include <limits.h>
+#include <errno.h>
+
+#define LINE_SIZE		16
+#define EBCDIC_SUBST_CHAR	(char)0x3F
+#define ASCII_SUBST_CHAR	(char)0x1A
+
+static void
+output_table (FILE *stream, const char *table)
+{
+	unsigned short i;
+	for (i = 0; i <= UCHAR_MAX; ++i) {
+		fprintf (stream, "%02X%c", (unsigned char)table[i],
+				 ((i + 1) % LINE_SIZE == 0) ? '\n' : ' ');
+	}
+}
+
+/* Build a pair of EBCDIC/ASCII translation tables using iconv */
+int
+gentable (FILE *stream, const char *code_ebcdic, const char *code_ascii)
+{
+	char ebcdic[UCHAR_MAX + 1], ascii[UCHAR_MAX + 1] = { 0 };
+	char *ebcdic_ptr = ebcdic, *ascii_ptr = ascii;
+	size_t ebcdic_size = UCHAR_MAX + 1, ascii_size = UCHAR_MAX + 1;
+	unsigned short i, nb_irreversible = 0;
+	iconv_t ic;
+
+	for (i = 0; i <= UCHAR_MAX; ++i) {
+		ebcdic[i] = (char)i;
+	}
+
+	ic = iconv_open (code_ascii, code_ebcdic);
+	if (ic == (iconv_t)-1) {
+		cb_error (_("conversion from %s to %s is not supported by your iconv implementation"), code_ascii, code_ebcdic);
+		return -1;
+	}
+
+	/* Note: POSIX iconv performs an implementation-defined conversion when
+	   encountering an untranslatable character (MUSL uses * and *BSD uses ?)
+	   and returns the number of such non-reversible conversions performed.
+	   GNU iconv instead sets errno to EILSEQ (unless //TRANSLIT is used).
+	   To cope with these differences, we convert the table character by
+	   character, and when encountering an untranslatable character, we
+	   map it to the ASCII substitution character (0x1A). This leaves a
+           number of unused ASCII characters: in the reverse translation table,
+           those are mapped to the EBCDIC substitution character (0x3F). */
+
+	for (i = 0; i <= UCHAR_MAX; ++i) {
+		size_t res;
+		ebcdic_size = 1;
+		res = iconv (ic, &ebcdic_ptr, &ebcdic_size, &ascii_ptr, &ascii_size);
+		if (res == (size_t)-1) {
+			if (errno == EILSEQ) {
+				/* GNU iconv: an untranslatable character was met */
+				*ascii_ptr = ASCII_SUBST_CHAR;
+				++ebcdic_ptr; ++ascii_ptr;
+				++ebcdic_size; ++ascii_size;
+				++nb_irreversible;
+			} else {
+				cb_error (_("an error occurred after converting %ld characters"), (ebcdic_ptr - ebcdic));
+				iconv_close (ic);
+				return -1;
+			}
+		/* POSIX iconv: a substitution was performed */
+		} else if (res != (size_t)0) {
```

Comment:
no need to cast zero; but we _may_ want to change the check to

```c
/* substitution necessary? */ 
if (res != 0) {
     if (res == (size_t)-1) {
     } else {
         /* POSIX iconv: a substitution was performed */
         ....
     }
}
```

Reply:
Yup, looks better, I'll do that.